DeprecationPage.md: Kokkos-3.7, 4.0 deprecation (#185)
kokkos/kokkos-core-wiki: docs/source/DeprecationPage.md

# Deprecation

> ## Kokkos-4.0 and Kokkos-3.7.0

Reviewer: It is not clear to me if these are the things marked for deprecation in 3.7 or 4.0 ... could we maybe have them in separate tables?

Reviewer 2: These are features we deprecated in 3.7.00. Hence, I read that as "between Kokkos 3.7.0 and Kokkos 4.0"

  | Deprecated Feature  | Replaced By          | Reason for Removal   |
  | --------------------| -------------------- | -------------------- |

>   | `Kokkos::is_reducer_type` | `Kokkos::is_reducer` | Streamline API

Reviewer: ```suggestion
  | `Kokkos::is_reducer_type` | `Kokkos::is_reducer` | Improve API
```

  | `OffsetView` constructors taking `index_list_type` | `std::pair (CPU), Kokkos::pair (GPU)` | Streamline arguments to `::pair` function

>   | overloads of `Kokkos::sort` taking a parameter `bool always_use_kokkos_sort` | Remove `always_use_kokkos_sort` bool parameter  | Updating overloads

Reviewer: We could stick to 'Improve API'

>   | Guard against non-public header inclusion | Core public headers: `Kokkos_core.cpp`, `Kokkos_Macros.hpp`, `Kokkos_Atomic.hpp`, `Kokkos_DetectionIdiom`, `Kokkos_MathematicalConstants.hpp`, `Kokkos_MathematicalFunctions.hpp`, `Kokkos_NumericTraits.hpp`, `Kokkos_Array.hpp`, `Kokkos_Complex.hpp`, `Kokkos_Pair.hpp`, `Kokkos_Half.hpp`, `Kokkos_Timer.hpp`; Algorithms public headers: `Kokkos_StdAlgorithms.hpp`, `Kokkos_Random.hpp`, `Kokkos_Sort.hpp`; Containers public headers: `Kokkos_Bit.hpp`, `Kokkos_DualView.hpp`, `Kokkos_DynRankView.hpp`, `Kokkos_ErrorReporter.hpp`, `Kokkos_Functional.hpp`, `Kokkos_OffsetView.hpp`, `Kokkos_ScatterView.hpp`, `Kokkos_StaticCrsGraph.hpp`, `Kokkos_UnorderedMap.hpp`, `Kokkos_Vector.hpp` | Improve API and reduce build time

Reviewer: ```suggestion
  | Guard against non-public header inclusion | Core public headers: `Kokkos_core.cpp`, `Kokkos_Macros.hpp`, `Kokkos_Atomic.hpp`, `Kokkos_DetectionIdiom.hpp`, `Kokkos_MathematicalConstants.hpp`, `Kokkos_MathematicalFunctions.hpp`, `Kokkos_NumericTraits.hpp`, `Kokkos_Array.hpp`, `Kokkos_Complex.hpp`, `Kokkos_Pair.hpp`, `Kokkos_Half.hpp`, `Kokkos_Timer.hpp`; Algorithms public headers: `Kokkos_StdAlgorithms.hpp`, `Kokkos_Random.hpp`, `Kokkos_Sort.hpp`; Containers public headers: `Kokkos_Bit.hpp`, `Kokkos_DualView.hpp`, `Kokkos_DynRankView.hpp`, `Kokkos_ErrorReporter.hpp`, `Kokkos_Functional.hpp`, `Kokkos_OffsetView.hpp`, `Kokkos_ScatterView.hpp`, `Kokkos_StaticCrsGraph.hpp`, `Kokkos_UnorderedMap.hpp`, `Kokkos_Vector.hpp` | Improve API and reduce build time
```

Reviewer: not sure if I would state, that this would reduce build time

  | Raise deprecation warnings if non-empty WorkTag class is used | Use empty WorkTag class | Improve API
  | `parallel_*` overloads taking the label as trailing argument | `Kokkos::parallel_*("KokkosViewLabel", policy, f);` | More logical ordering of parameters
  | Embedded types (`argument_type`, `first_argument_type`, and `second_argument_type`) in `std::function` | types removed in C++20 | Deprecation in `Kokkos_Functional.hpp` mirrors that in `std::function` (`#include <functional>`) 

>   | `InitArguments` struct | `InitializationSettings()` object with attributes that can be queried | Align with object-oriented programming

Reviewer: ```suggestion
  | `InitArguments` struct | `InitializationSettings` object with attributes that can be queried | Make initialization more transparent and easier to understand
```

  | `finalize_all()` | `finalize_spaces()`| Improve  API

>   | Command-line arguments (other than `--help`) that are not prefixed with `kokkos-*` | `--kokkos-num-threads`, `--kokkos-device-id`, `--kokkos-num-devices`| Improve API

Reviewer: ```suggestion
  | Command-line arguments (other than `--help`) that are not prefixed with `--kokkos-` | `--kokkos-num-threads`, `--kokkos-device-id`, `--kokkos-num-devices`| Improve API
```

# Deprecation

## Kokkos-3.7.01 - 4.0*

  | Deprecated Feature  | Replacement          | Reason for Deprecating   |
  | --------------------| -------------------- | ------------------------ |
  | `Kokkos::is_reducer_type` | `Kokkos::is_reducer` | Improve API
  | `OffsetView` constructors taking `index_list_type` | `Kokkos::pair` (CPU and GPU) | Streamline arguments to `::pair` function
  | overloads of `Kokkos::sort` taking a parameter `bool always_use_kokkos_sort` | Use `Kokkos::BinSort` if required, or call `Kokkos::sort` without bool parameter  | Updating overloads

>   | Guard against non-public header inclusion | Core public headers: `Kokkos_core.cpp`, `Kokkos_Macros.hpp`, `Kokkos_Atomic.hpp`, `Kokkos_DetectionIdiom`, `Kokkos_MathematicalConstants.hpp`, `Kokkos_MathematicalFunctions.hpp`, `Kokkos_NumericTraits.hpp`, `Kokkos_Array.hpp`, `Kokkos_Complex.hpp`, `Kokkos_Pair.hpp`, `Kokkos_Half.hpp`, `Kokkos_Timer.hpp`; Algorithms public headers: `Kokkos_StdAlgorithms.hpp`, `Kokkos_Random.hpp`, `Kokkos_Sort.hpp`; Containers public headers: `Kokkos_Bit.hpp`, `Kokkos_DualView.hpp`, `Kokkos_DynRankView.hpp`, `Kokkos_ErrorReporter.hpp`, `Kokkos_Functional.hpp`, `Kokkos_OffsetView.hpp`, `Kokkos_ScatterView.hpp`, `Kokkos_StaticCrsGraph.hpp`, `Kokkos_UnorderedMap.hpp`, `Kokkos_Vector.hpp` | Improve API

Reviewer 2: ```suggestion
  | Guard against non-public header inclusion | Core public headers: `Kokkos_Core.hpp`, `Kokkos_Macros.hpp`, `Kokkos_Atomic.hpp`, `Kokkos_DetectionIdiom.hpp`, `Kokkos_MathematicalConstants.hpp`, `Kokkos_MathematicalFunctions.hpp`, `Kokkos_NumericTraits.hpp`, `Kokkos_Array.hpp`, `Kokkos_Complex.hpp`, `Kokkos_Pair.hpp`, `Kokkos_Half.hpp`, `Kokkos_Timer.hpp`; Algorithms public headers: `Kokkos_StdAlgorithms.hpp`, `Kokkos_Random.hpp`, `Kokkos_Sort.hpp`; Containers public headers: `Kokkos_Bitset.hpp`, `Kokkos_DualView.hpp`, `Kokkos_DynRankView.hpp`, `Kokkos_ErrorReporter.hpp`, `Kokkos_Functional.hpp`, `Kokkos_OffsetView.hpp`, `Kokkos_ScatterView.hpp`, `Kokkos_StaticCrsGraph.hpp`, `Kokkos_UnorderedMap.hpp`, `Kokkos_Vector.hpp` | Improve API
```

  | Raise deprecation warnings if non-empty WorkTag class is used | Use empty WorkTag class | Improve API

>   | `parallel_*` overloads taking the label as trailing argument | `Kokkos::parallel_*("KokkosViewLabel", policy, f);` | Consistent ordering of parameters

Reviewer 3: ```suggestion
  | `parallel_*` overloads taking the label as trailing argument | `Kokkos::parallel_*("KokkosKernelLabel", policy, f);` | Consistent ordering of parameters
```

  | Embedded types (`argument_type`, `first_argument_type`, and `second_argument_type`) in `std::function` | Use `decltype` if required | Deprecation in `Kokkos_Functional.hpp` mirrors that in `std::function` (`#include <functional>`) 
  | `InitArguments` struct | `InitializationSettings()` object with attributes that can be queried | Make initialization transparent and understandable
  | `finalize_all()` | `finalize()`| Improve  API
  | Command-line arguments (other than `--help`) that are not prefixed with `kokkos-*` | `--kokkos-num-threads`, `--kokkos-device-id`, `--kokkos-num-devices`| Improve API

>   | `--[kokkos-]numa` command-line argument and `KOKKOS_NUMA` environment variable | `--kokkos-num-threads`| Harmonize option nomenclature with that of C++ `std::thread` library

Reviewer 3: Should the replacement be `numa` something or other? I'm pretty sure it's not `num-threads`
